ludwig/decoders/sequence_decoders.py

```diff
@@ -13,18 +13,227 @@
 # limitations under the License.
 # ==============================================================================
 import logging
+from typing import Dict, Tuple
 
-from ludwig.constants import SEQUENCE, TEXT
+import torch
+import torch.nn as nn
+
+from ludwig.constants import HIDDEN, LOGITS, SEQUENCE, TEXT
 from ludwig.decoders.base import Decoder
 from ludwig.decoders.registry import register_decoder
+from ludwig.utils import strings_utils
 
 logger = logging.getLogger(__name__)
 
 
-# TODO(Justin): Implement sequence decoders.
+class RNNDecoder(nn.Module):
+    """GRU or RNN-based decoder."""
+
+    def __init__(self, hidden_size: int, vocab_size: int, cell_type: str):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.vocab_size = vocab_size
+        self.embedding = nn.Embedding(vocab_size, hidden_size)
+        self.relu = nn.ReLU()
+        if cell_type == "gru":
+            self.rnn = nn.GRU(hidden_size, hidden_size, batch_first=True)
+        else:
+            self.rnn = nn.RNN(hidden_size, hidden_size, batch_first=True)
+        self.out = nn.Linear(hidden_size, vocab_size)
+
+    def forward(self, input: torch.Tensor, hidden: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Runs a single decoding time step.
+
+        Modeled off of https://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html.
+
+        Args:
+            input: [batch_size] tensor with the previous step's predicted symbol.
+            hidden: [batch_size, hidden_size] tensor with the previous step's hidden state.
+
+        Returns:
+            Tuple of two tensors:
+            - output: [batch_size, vocab_size] tensor with the logits.
+            - hidden: [batch_size, hidden_size] tensor with the hidden state for the next time step.
+        """
+        # Unsqueeze predicted tokens.
+        input = input.unsqueeze(1).to(torch.int)
+        output = self.relu(self.embedding(input))
+        output, hidden = self.rnn(output, hidden)
+        output_logits = self.out(output)
+        return output_logits, hidden
+
+
+class LSTMDecoder(nn.Module):
+    """LSTM-based decoder."""
+
+    def __init__(self, hidden_size: int, vocab_size: int):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.vocab_size = vocab_size
+        self.embedding = nn.Embedding(vocab_size, hidden_size)
+        self.relu = nn.ReLU()
+        self.lstm = nn.LSTM(hidden_size, hidden_size, batch_first=True)
+        self.out = nn.Linear(hidden_size, vocab_size)
```

Comment:
same thing for embedding and out linear optionally using the same weights as for the rundecoder

Reply:
Done.